### ISSUE TITLE
Meta - API Endpoint for operations

### ISSUE THREAD

**Issue by juherr** [steve-community/steve#1000]
This feature request completes #990 and could be a meta ticket for all tickets for a dedicated operation.

Could be done in steps:
1. `Remote Start` / `Remote Stop` (#990)
2. The ones provided by [OCPI](https://github.com/ocpi/ocpi/blob/release-2.2.1-bugfixes/mod_commands.asciidoc)
2.1.  `Reserve Now` + `Cancel Reservation`
2.2. `Unlock Connector`
3. Everything else
3.1. `Change Availability`
3.2. `Get Configuration` / `Change Configuration`
3.3. `Clear Cache`
3.4. `Get Diagnostics`
3.5. `Reset`
3.6. `Update Firmware`
3.7. `Data Transfer`
3.8. `Get Local List Version` / `Send Local List`
3.9. `Trigger Message`
3.10. `Get Composite Schedule`
3.11. `Clear Charging Profile` / `SetCharging Profile`

**Comment by srigurubyo** [comment]
Would be really super useful!

**Comment by alexandre-da-costa** [comment]
In my opinion, step 1 is essential to integrate with third party services.

I am actually in dire need of it right now to be able to advance in a university research, where I have to able to remote start and stop charging operations from my backend.

I am a PHP guy, and have little experiencie with Java, but will try to work on these endpoints. 

Or does someone already have something working and can help a brother? 

Thanks.

**Comment by NikSays** [comment]
Hello,
A while back our team actually implemented some REST endpoints ourselves. Current implementation can be seen [here](https://github.com/NikSays/steve/tree/60db0c039282ec321f4235d2c8e7b59b8c8e009a/src/main/java/de/rwth/idsg/steve/web/controller/rest). It does not comply to the code style of this repository, but we are interested in fixing it up and merging it here. 

Firstly, we can help with step 1 (`Remote start/stop`)

Then, we propose some other endpoints
- `Get chargepoint list`
- `Get transaction details` (including meter values)

We consider those to be important for 3rd party integration  

Thanks.

**Comment by juherr** [comment]
@NikSays Thanks for the idea. I updated the issue description accordingly.
I'm not the owner of the project but I think you can go ahead and propose some pull requests with the changes you proposed.

**Comment by goekay** [comment]
i slightly remember that i expressed my general opinion about this in some other issue, but cannot find it at the moment. here is a short version:

i am quite wary about exposing these as APIs... not much the data-related ones, but the ocpp operations. here is the reason: every feature request comes from a specific perspective and use case, and addresses %1 of the functionality. this is easy to do in a fork that solves _your_ problem, but in a mainstream project like this one i have to think about the completeness and other %99. step by step, if completeness is desired, it will arrive at the same functionality and complexity as ocpp. some operations even have some slight differences between versions. so, you have to respect ocpp versions as well. great, we recreate ocpp but in REST form.

**Comment by juherr** [comment]
@goekay I understand your "ocpp first" approach and you are right when you say that the API will look like OCPP at the end.

But as we have seen many times, users expect a simpler way to deal with stations which is not OCPP (at least because they don't want manage the different OCPP transports). https://github.com/steve-community/steve/issues/117 is one uses-case between some which will benefit of an API.
Steve is an awesome piece of software that is doing all the technical OCPP parts.
At the same time, I don't think OCPP explains how to deal with the CentralPoint from the outside.
OCPI is maybe the closest approach to that (#56) but doesn't cover (yet?) all operations (at least in the official modules) and has its own complexity.

For reference, some API descriptions by similar softwares:
- https://developer.greenflux.com/docs/remote-commands
- https://api.emabler.net/api/swagger/ui

On another side, the current UI of Steve is working well but is a bit raw. Having a full feature API will open the opportunity to create external projects dedicated to UI, with maybe another UX approachs, and trying to fix its own problems (for example #991 or #1029).

If I read well between the lines, you propose to fork the project but I don't think it will benefit the community and the effort would be large to keep the sync with the upstream.

Please, could you specify your thinking?

**Comment by engsandro** [comment]
Hi guys, I have some GreenFlux Smart Charging Controller DSCU (discontinued products) that I need to update the firmware to the latest available version, probably 4.3.5). Does anyone know the server address to download the FW?

[https://www.eautotoltokabel.hu/wp-content/uploads/2021/04/GreenFlux-DSCU_Smart-Charging-Controller-installation-manual-v4.6.pdf](https://github.com/debabrata-das/GreenFlux/discussions/url)

Thanks

**Comment by goekay** [comment]
@juherr 

> If I read well between the lines, you propose to fork the project...

not really. in fact, i would really would like to prevent that. i do not propose a solution, i don't have a solution. i am just raising a concern and trying to engage the community in active decision making. my concern is a parallel ocpp-like behaviour with different technical realisation and the maintenance hell that would come with it.

> On another side, the current UI of Steve is working well but is a bit raw. 

i would agree. i am not a frontend developer. i tried my best. PRs and contributions are always welcome in that area. but there was nothing so far in the last X years.

**Comment by emerout** [comment]
@juherr 
> On another side, the current UI of Steve is working well but is a bit raw. Having a full feature API will open the opportunity to create external projects dedicated to UI, with maybe another UX approachs, and trying to fix its own problems (for example #991 or #1029).

This pull request https://github.com/steve-community/steve/pull/1074 could be useful for creating a different UI without implementing a brand new API.

**Comment by mtomoda** [comment]
Thx @juherr 
But I cant understand.

I edited main.properties and rerun maven.

For example, how should the API be written to start charging from client?

**Comment by juherr** [comment]
@mtomoda As you can see in the issue description, the api you expect doesn't exist for the moment.

**Comment by vranbat** [comment]
I'm also very interested in the possibility of having an API for RemoteStart and RemoteStop. 
It would be very useful for me to be able to set the charging current as well. I would like to manage the charging based on the availability of solar power using a Python script. 
Thank you very much for the great work you have done!

**Comment by sijojamesjohn** [comment]
How to use API? AUTH?

**Comment by imre-evo** [comment]
Hi Everybody!
Are there any updates in the API endpoints? I can see lots of new features idea, for example start and stop transaction or read the meter values during a transaction, but I don't find them in the manager/v3/api-docs. Are these endpoints already implemented, or is it in the future plans?
Thanks for the answer in advance!

**Comment by smeatond** [comment]
Is there any updates on this?

**Comment by RealRunSpace** [comment]
Bumping this, I'd love to even sponsor the project if we get the full API!

**Comment by juherr** [comment]
Hi @goekay, congratulations on the launch of Powerfill. The work done on the UI and the Steve integration is genuinely impressive.

Since you've implemented APIs covering most of the expected features, I'm curious about the long-term plan regarding this topic. Do you intend to backport these APIs into Steve, or do you see them as a key differentiator for Powerfill, in which case upstream contributions would not be considered?

Thanks in advance for the clarification.

**Comment by RealRunSpace** [comment]
To be honest I actually created my "own way" to have an API, a .php file that when called will directly perform requests in the Steve backend and return data once the server process it, but I really wish we get a proper API to manage all these things directly over Steve!

**Comment by goekay** [comment]
hey @juherr, thanks for the kind words! 

> Do you intend to backport these APIs into Steve, or do you see them as a key differentiator for Powerfill, in which case upstream contributions would not be considered?

and, thanks for raising this. it’s a fair question. how i imagine and plan Powerfill's API contributions to SteVe:

Powerfill will contribute back to SteVe in _phases_. early on, the newest capabilities, especially those with strong business relevance, may live as closed source while they are actively evolving. as things mature and we move on to the next set of ideas, earlier work will be open sourced and contributed back after a natural "decay" period.

our intent here is to build a mutually beneficial relationship between Powerfill and the community of SteVe: some of the features will be developed completely open-source and will benefit everyone (incl. Powerfill), while some will be developed within Powerfill initially and then _donated_ to SteVe after a while. importantly, those contributions will not be limited to APIs alone.

regarding your emphasis on consideration of upstream contributions: they will be considered. for sure. they are welcome even. but contributors should keep in mind, that these might be used by everyone... non-commercially and commercially (for ex. by Powerfill). this was always the case so far anyways. therefore, nothing changes there.

the goal is to balance fast innovation and sustainability on our side of Powerfill, while ensuring the community benefits in a growing, long term way as well.

these are just some rough ideas. we don't have a concrete timing or roadmap regarding this. hopefully, things will evolve naturally. but... watch this space, because something will happen in the next few days ;)

--

on another note, the external communications about the launch of Powerfill platform did not happen yet. they will in the next few days though. for the impatient: if anyone wants to join, i am inviting you to come to play with us at https://powerfill.app/register

**Comment by juherr** [comment]
I assumed the announcement was already public following the emails sent yesterday. Apologies if I unintentionally shared something that was not meant to be public.

I'd like to better understand your position regarding the APIs. Given that the implementation already exists, it's unclear to me whether upstream API work would still make sense at this stage.

Could you give some sign of the expected timeline for API contributions to Steve? Are we thinking in terms of months, or longer-term?
